8Knot/pages/codebase/visualizations/cntrb_file_heatmap.py

```diff
@@ -0,0 +1,385 @@
+from dash import html, dcc, callback
+import dash
+import dash_bootstrap_components as dbc
+import dash_mantine_components as dmc
+from dash.dependencies import Input, Output, State
+import plotly.graph_objects as go
+import pandas as pd
+import logging
+from dateutil.relativedelta import *  # type: ignore
+import plotly.express as px
+from pages.utils.graph_utils import get_graph_time_values, color_seq
+from queries.contributors_query import contributors_query as cnq
+from queries.cntrb_per_file_query import cntrb_per_file_query as cpfq
+from queries.repo_files_query import repo_files_query as rfq
+from app import augur
+import io
+from cache_manager.cache_manager import CacheManager as cm
+from pages.utils.job_utils import nodata_graph
+import time
+from dash.exceptions import PreventUpdate
+
+PAGE = "codebase"
+VIZ_ID = "cntrb-file-heatmap"
+
+# div to hold all objects to wait for loading to render
+graph_loading = html.Div(
+    [
+        dbc.Popover(
+            [
+                dbc.PopoverHeader("Graph Info:"),
+                dbc.PopoverBody(
+                    """
+                    This visualization takes all of the contributors that have ever made a pull request\n
+                    including the specific directory or file and plots the last time they have made\n
+                    a contribution in any part of the codebase. See the definition of "contribution"\n
+                    on the Info page for more information.
```

Comment:
Suggestion:

"This visualization helps to analyze the activity of the past contributors to sub-sections (files or folders) of a repository's codebase. Specifically, this heatmap identifies the time-period in which a sub-section's contributors (those people who have opened at least one pull request to contribute to a sub-section) last opened a PR for files anywhere in the codebase. This could be interpreted as monitoring technical knowledge retention of codebase components: if a sub-section's past contributors are no longer code-contributors, maintainership of that sub-section could be insufficient and require attention."

This adds an interpretation, which we've historically avoided doing, but I think for this visualization it's necessary.

Reply:
@JamesKunstle So this portion of the description "last opened a PR for files anywhere in the codebase" is actually not correct. The "last activity" is not limited to pull_requests, it is any "contribution" (in our definition) which includes issues prs comments reviews etc etc. This would be my suggested version of what you have put: 

"This visualization analyzes the activity of the contributors to sub-sections (files or folders) of a repository. Specifically, this heatmap identifies the last time a sub-section's contributors (those people who have opened at least one pull request to a sub-section) last contributed to the repository. See the definition of "contribution" on the Info page for more information. This could be interpreted as monitoring technical knowledge retention of codebase components: if a sub-section's past contributors are no longer active in the repository, maintainership of that sub-section could be insufficient and require attention."

Reply:
Oh I see, we're using contributor_actions not PRs specifically. Your update 100% tracks, love it.